### PR TITLE
Allow remappings to change depending on the context.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: docs/layout-of-source-files.rst [hunk 1]
```diff
@@ -68,15 +70,20 @@ remappings so that e.g. ``github.com/ethereum/dapp-bin/library`` is remapped to
 ``/usr/local/dapp-bin/library`` and the compiler will read the files from there. If
 remapping keys are prefixes of each other, the longest is tried first. This
 allows for a "fallback-remapping" with e.g. ``""`` maps to
-``"/usr/local/include/solidity"``.
+``"/usr/local/include/solidity"``. Furthermore, these remappings can be
```
Comment (redsquirrel) [inline docs/layout-of-source-files.rst]: "can be depend on the context"

**Comment by VoR0220** [comment]
got any ideas for how to put some tests together for this?

**Comment by raineorshine** [comment]
I like that this allows for multiple versions of libraries to work side-by-side. That seems essential for a higher level package management system to sit on top of. I wonder if doing this at the import level is the best place though. It would be _very_ confusing if a user used two identical import statements and one of them works as expected and the other errors out. I suppose that the package manager/module loader is expected to prevent that from happening? There must be a similar problem that was solved in npm to allow different library versions to be used in the same project. A little different perhaps since Javascript does not require linking them all together with single compile command.

**Comment by chriseth** [comment]
@raineorshine I am not really sure if the situation you describe would be confusing or not. Assume you have a package whose package description tells the packager that it wants to use the `strings` library as of version 1.0.5. Then another package wants to use `strings` as of version 1.6.2. Both use `import "strings";` and they end up importing different files. On the contrary: Wouldn't it be confusing if a remapping configuration for one package would be able to modify the remappings for another one?

**Comment by chriseth** [comment]
I just changed this commit so that the remappings are a feature of the compiler. This allows better post-compilation verification and does not require a change to the js interface at the same time.
A future change will replace the two options we currently have (optimizer and number of runs) by a much more detailed options json object which will also include the remappings.

**Comment by chriseth** [comment]
@VoR0220 moving them into the compiler stack also allows us to test them more easily :-)

**Comment by raineorshine** [comment]
@chriseth That makes sense. I guess I just get scared when I hear "context dependent"... I generally work in environments where referential transparency is a good thing. So this is just different than what I am used to.

**Pull request merged**